### PR TITLE
Append history in interactive sessions

### PR DESCRIPTION
**Pull request by ylluminarious**
As per the discussion [here.](https://trac.macports.org/ticket/56381) Please let me know if you have any improvements in mind. Pinging @raimue since we spoke about this together on the aforementioned Trac ticket.

### CODE REVIEW

File: src/pextlib1.0/readline.c [hunk 1]
```diff
@@ -309,6 +311,16 @@ int RLHistoryCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int objc, Tcl
...
+	}  else if (0 == strcmp("append", action)) {
+		if (objc != 3) {
+			Tcl_WrongNumArgs(interp, 1, objv, "append filename");
+			return TCL_ERROR;
+		}
+		s = Tcl_GetString(objv[2]);
+
+		hist_file = fopen(s, "a");
+		fprintf(hist_file, "%s\n", current_history()->line);
+		fclose(hist_file);
```
Comment (raimue) [inline src/pextlib1.0/readline.c]: There is a chance for a race condition here when multiple port instances are running at the same time. Although it is very unlikely that anyone would be able to type in both port instances at the same time, so we could just ignore it.

Comment (ylluminarious) [reply src/pextlib1.0/readline.c]: @raimue Hmm, okay, I will see if I can address this with lockfiles or the like. If things get too hairy, I may just ignore it as you say.

**Comment by ylluminarious** [comment]
@raimue 

> I am sorry I did not review your patch on Trac. My time for MacPorts is currently a bit scarce, but the pull request makes it easier. Thank you for that.

No problem, I figured you were probably too busy to respond on Trac. I definitely understand not having time for things like Macports sometimes; real life can be a pain! But you're welcome, and thank you for the kindness.

> The changes look good to me, although I did not test them. I would appreciate if you could use the usual trailer with "Closes: ..." as reference to the ticket.

Okay, I will try to do that in the future. Where am I supposed to put the "Closes: ..." remark? In the title of the PR?

**Comment by jmroot** [comment]
> Okay, I will try to do that in the future. Where am I supposed to put the "Closes: ..." remark? In the title of the PR?

In the commit message body, on a line by itself.
https://trac.macports.org/wiki/CommitMessages

**Comment by ylluminarious** [comment]
@jmroot Oh, okay. I had read that article previously, but I missed the part which gives an example of that. Thanks for pointing me to it!

**Comment by pmetzger** [comment]
@ylluminarious You will want to squash your commits into a single commit...

**Comment by ylluminarious** [comment]
@raimue I have added some file-locking code as discussed above. Sorry for the delay; I've been quite busy but finally managed to make some time for this. Please let me know if anything else should be done. Also, should I squash my commits into a single one to have the "Closes: ..." trailer?

**Comment by pmetzger** [comment]
You must indeed always squash your commits. (I have no opinion on the PR itself...)

**Comment by ylluminarious** [comment]
@pmetzger Ok, thanks for the info.

**Comment by ylluminarious** [comment]
@raimue Friendly ping... any thoughts on this? :)

File: src/pextlib1.0/readline.c [hunk 2]
```diff
@@ -309,6 +311,16 @@ int RLHistoryCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int objc, Tcl
...
 		s = Tcl_GetString(objv[2]);
 		write_history(s);
+	}  else if (0 == strcmp("append", action)) {
+		if (objc != 3) {
+			Tcl_WrongNumArgs(interp, 1, objv, "append filename");
+			return TCL_ERROR;
+		}
+		s = Tcl_GetString(objv[2]);
+
+		hist_file = fopen(s, "a");
```
Comment (neverpanic) [inline src/pextlib1.0/readline.c]: Can you please add error handling to the `fopen(3)` call here? `mkdtemp.c` has an example on how to do this, e.g.

```c
if (NULL == (hist_file = fopen(s, "a"))) {
    Tcl_AppendResult(interp, "fopen(", s, "): ", strerror(errno), NULL);
    return TCL_ERROR;
}
```

File: src/pextlib1.0/readline.c [hunk 3]
```diff
@@ -309,6 +315,19 @@ int RLHistoryCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int objc, Tcl
 		}
 		s = Tcl_GetString(objv[2]);
 		write_history(s);
+	}  else if (0 == strcmp("append", action)) {
+		if (objc != 3) {
+			Tcl_WrongNumArgs(interp, 1, objv, "append filename");
+			return TCL_ERROR;
+		}
+		s = Tcl_GetString(objv[2]);
+		hist_file = fopen(s, "a");
```
Comment (neverpanic) [inline src/pextlib1.0/readline.c]: Thank you GitHub review for hiding my review comment against the version in one of the lower commits. Please add error handling here. See `mkdtemp.c` for good examples. E.g., use:

```tcl
if (NULL == (hist_file = fopen(s, "a"))) {
    Tcl_AppendResult(interp, "fopen(", s, "): ", strerror(errno), NULL);
    return TCL_ERROR;
}
```

File: src/pextlib1.0/readline.c [hunk 4]
```diff
@@ -317,9 +321,12 @@ int RLHistoryCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int objc, Tcl
 			return TCL_ERROR;
 		}
 		s = Tcl_GetString(objv[2]);
-
 		hist_file = fopen(s, "a");
+
+		flock(fileno(hist_file), LOCK_EX);
 		fprintf(hist_file, "%s\n", current_history()->line);
```
[the remark this answers is not in the document]
Comment (ylluminarious) [reply src/pextlib1.0/readline.c]: @neverpanic Ok, I went ahead and took your advice for `fflush`. Strangely, what you're saying makes perfect sense, but when I tested this out, the locking mechanism actually worked. It was as though it didn't need a flush. I tested it by opening two separate terminals and entering commands into them via [CloneKeys.](https://code.google.com/archive/p/clonekeys/) I'm curious now; do you know why that may have happened?

File: src/pextlib1.0/readline.c (hunk 3 above)
Comment (ylluminarious) [reply src/pextlib1.0/readline.c]: @neverpanic Thanks for the code, I went ahead and pasted it in verbatim :)

I also added some error-checking for the `fprintf` call; please let me know if it's okay. It's not as pretty as the code in `mktemp.c`, but it should do the trick.

**Comment by ylluminarious** [comment]
@neverpanic I went ahead and implemented your changes as you requested. Please let me know if they're satisfactory. I also added some comments for visual and conceptual clarity since I stared at this code without comments and it felt like my eyes were getting poked with curly-braced needles :)

**Comment by neverpanic** [comment]
Thanks!

**Comment by ylluminarious** [comment]
@neverpanic You're welcome! Thanks for merging it and for the help!

Did you see [my comment](https://github.com/macports/macports-base/pull/99#discussion_r211449539) from last night wherein I had a question about the flushing seeming unnecessary? I'm curious to know what may have been going on... :)

**Comment by ylluminarious** [comment]
@neverpanic Thanks so much for the explanation! This makes more sense to me now. I suspected that `fclose` was behind this. I guess the lock gave just enough of a delay that the flushing done by `fclose` happens right before the acquisition of the lock for the competing process, or at least before *that* competing process was able to do *its* flush. Thus, there was a *de facto* order of operations, although unreliable.

Yes, such a situation would indeed make it hard to expose the problem. That is a very hard-to-hit race condition, although possible. And as you also said, if the two writes were mixed into the same syscall then there would be no conflict anyway. Maybe I'll try that test you mention someday, just for fun :D

**Pull request merged**